### PR TITLE
Expand channels to allow them to contain environment variables

### PR DESCRIPTION
**Pull request by mariusvniekerk**
These environment variables can get templated later at render time and are useful for things like usernames, passwords and tokens.

This enables the use-case where a channel is backed by Anaconda Enterprise or Quetz

### CODE REVIEW

**Comment by mariusvniekerk** [comment]
@wolfv does micromamba not support channels with ports?

**Comment by wolfv** [comment]
@mariusvniekerk do you have an example? Should be supported but there can always linger a bug!

**Comment by mariusvniekerk** [comment]
The test failure here for micromamba is the example.

File: docs/authenticated_channels.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+--channel 'http://host.com/t/$MY_REPO_TOKEN/channel'
+```
+
+If you accidentally pass a channel url that contains a token like so
+
+```sh
+--channel "http://host.com/t/$MY_REPO_TOKEN/channel"
+```
+
+Then conda lock will detect the environment variabl used (provided that the environment variable used ends in a sensible suffix (KEY, TOKEN, PASS, etc)).
```
Comment (cpcloud) [inline docs/authenticated_channels.md:24]: This seems too magical. Is anything lost by not having this magic?

File: docs/authenticated_channels.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+The _name_ of the environment variable will form part of your lock and you will have to have that SAME environment variable set if you wish to run the install
+
+```sh
+# retrieve secrets from some store
+source $(./get-lockfile-env-vars-from-secret-store)
+# use the secrets as part of the conda-lock invocation
+conda-lock install -n my-env-with-a-secret condae-lock.yml
```
Comment (cpcloud) [inline docs/authenticated_channels.md:32]:
```suggestion
conda-lock install -n my-env-with-a-secret conda-lock.yml
```

File: docs/authenticated_channels.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+## Simple Auth
+
+For other channels (such as those self-managed) you may be using a more standard username/password auth
```
Comment (cpcloud) [inline docs/authenticated_channels.md:37]:
```suggestion
For other channels (such as those self-managed) basic auth is supported
```

File: conda_lock/models/test_channel.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    monkeypatch.setenv("PASSWORD", passwd)
+
+    # These two urls are equivalent since we can pull the env vars out.
+    x = env_var_normalize("http://$USER:$PASSWORD@host/prefix/t/$TOKEN/suffix")
+    y = env_var_normalize(f"http://{user}:{passwd}@host/prefix/t/{token}/suffix")
+
+    assert x.env_var_url == y.env_var_url
+
+    replaced = y.conda_token_replaced_url()
+    assert replaced == f"http://{user}:{passwd}@host/t/<TOKEN>/prefix/suffix"
```
Comment (maresb) [inline conda_lock/models/test_channel.py:43]: @mariusvniekerk, I'm trying to understand this code, and this looks wrong to me. Is there some crazy reason why `/prefix/` comes after and not before `<TOKEN>` in this assert?

**Pull request merged**